drgn_tools/dm.py

```diff
@@ -0,0 +1,118 @@
+# Copyright (c) 2023, Oracle and/or its affiliates.
+# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/
+"""
+Helpers for device mapper devices.
+"""
+import argparse
+from typing import Iterable
+from typing import Tuple
+
+from drgn import Object
+from drgn import Program
+from drgn.helpers.linux.list import list_for_each_entry
+from drgn.helpers.linux.rbtree import rbtree_inorder_for_each_entry
+
+from drgn_tools.corelens import CorelensModule
+from drgn_tools.module import ensure_debuginfo
+from drgn_tools.table import print_table
+from drgn_tools.util import BitNumberFlags
+from drgn_tools.util import kernel_version
+from drgn_tools.util import variable_exists
+
+
+def for_each_dm_hash(prog: Program) -> Iterable[Tuple[Object, str]]:
+    # bucket size is 64
+    for i in range(64):
+        head = prog["_name_buckets"][i]
+        for hc in list_for_each_entry(
+            "struct hash_cell", head.address_of_(), "name_list"
+        ):
+            yield hc.md, hc.name.string_().decode()
+
+
+def for_each_dm_rbtree(prog: Program) -> Iterable[Tuple[Object, str]]:
+    for hc in rbtree_inorder_for_each_entry(
+        "struct hash_cell", prog["name_rb_tree"], "name_node"
+    ):
+        yield hc.md, hc.name.string_().decode()
+
+
+def for_each_dm(prog: Program) -> Iterable[Tuple[Object, str]]:
+    if variable_exists(prog, "_name_buckets"):
+        dm_list = list(for_each_dm_hash(prog))
+    elif variable_exists(prog, "name_rb_tree"):
+        dm_list = list(for_each_dm_rbtree(prog))
+
+    for dm, name in dm_list:
+        yield dm, name
```

Comment:
You're creating the full list here, and then returning an iterator over that list. It doesn't really make sense to collect everything using `list()` unless you plan on returning the list directly. I would drop the `list()` call and directly return the iterables:
```suggestion
def for_each_dm(prog: Program) -> Iterable[Tuple[Object, str]]:
    if "_name_buckets" in prog:
        return for_each_dm_hash(prog)
    elif "name_rb_tree" in prog:
        return for_each_dm_rbtree(prog)
    else:
        raise NotImplementedError("Cannot find dm devices")
```

But if you want to return a fully-collected list, you could do that too:

```python
def for_each_dm(prog: Program) -> List[Tuple[Object, str]]:
    if "_name_buckets" in prog:
        return list(for_each_dm_hash(prog))
    elif "name_rb_tree" in prog:
        return list(for_each_dm_rbtree(prog))
    else:
        raise NotImplementedError("Cannot find dm devices")
```